### ISSUE TITLE
Support Symfony\Process >= 4.0

### ISSUE THREAD

**Issue by repat** [willdurand/nmap#28]
Otherwise it's not possible to use it with Laravel 5.6

**Comment by m7mdcc** [comment]
@repat Thanks for this : https://github.com/repat/nmap

**Comment by repat** [comment]
Yea but I didn't commit/push my tests yet. Have a look now :) I basically just removed escapeArguments because you can now escape the using the constructor afaik. I didn't get to that yet

**Comment by willdurand** [comment]
Can someone try the `master` branch?

**Comment by m7mdcc** [comment]
@willdurand  it doesn't show mentioned error above ,  but when i try to : 
$hosts = \Nmap\Nmap::create()->enableVerbose()->scan([ '192.168.10.1' ]); 

its return empty array .

**Comment by repat** [comment]
Which PHP Version do you use @m7mdcc?
```
$php --version
PHP 7.2.2-3+ubuntu16.04.1+deb.sury.org+1 (cli) (built: Feb  6 2018 16:11:23) ( NTS )
```

---

I tried it every way possible. `php artisan tinker` is run with `root`, so that might explain the differences.

tl;dr:
* Both `enableServiceInfo()` and `enableOsDetection()` fail (504 Timeout/`Symfony\Component\Process\Exception\ProcessTimedOutException` and `root` needed)
* Without any parameters or with `enableVerbose()` either 504 Timeout or empty array

---

```php
$hosts = \Nmap\Nmap::create()->enableServiceInfo()->scan(['repat.de']);
```

In Laravel via the browser in a TestController:

> 504 Gateway Time-out
> nginx/1.10.3 (Ubuntu)

Via `php artisan tinker`:

> Symfony\Component\Process\Exception\ProcessTimedOutException with message 'The process "nmap -sV -oX '/tmp/nmap-scan-output.xmlKtkkLe' 'repat.de'" exceeded the timeout of 60 seconds.'`

---

```php
$hosts = \Nmap\Nmap::create()->scan(['repat.de']);
```

In Laravel via the browser in a TestController:
> 504 Gateway Time-out
> nginx/1.10.3 (Ubuntu)

or sometimes:

```json
[
  {
    
  }
]
```

However via `php artisan tinker`:

```
Array
(
    [0] => Nmap\Host Object
        (
            [addresses:Nmap\Host:private] => Array
                (
                    [95.143.172.170] => Nmap\Address Object
                        (
                            [address:Nmap\Address:private] => 95.143.172.170
                            [type:Nmap\Address:private] => ipv4
                            [vendor:Nmap\Address:private] =>
                        )

                )

            [state:Nmap\Host:private] => up
            [hostnames:Nmap\Host:private] => Array
                (
                    [0] => Nmap\Hostname Object
                        (
                            [name:Nmap\Hostname:private] => repat.de
                            [type:Nmap\Hostname:private] => user
                        )

                    [1] => Nmap\Hostname Object
                        (
                            [name:Nmap\Hostname:private] => aries.uberspace.de
                            [type:Nmap\Hostname:private] => PTR
                        )

                )

            [ports:Nmap\Host:private] => Array
                (
                    [0] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 22
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => ssh
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [1] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 80
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => http
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [2] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 110
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => pop3
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [3] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 113
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => ident
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [4] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 143
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => imap
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [5] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 443
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => https
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [6] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 587
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => submission
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [7] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 993
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => imaps
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [8] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 995
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => pop3s
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                )

        )

)
=> true
```

---

```php
$hosts = \Nmap\Nmap::create()->enableOsDetection()->scan(['repat.de']);
```

In Laravel via the browser in a TestController:

![image](https://user-images.githubusercontent.com/516807/37442567-b00cacd6-27cc-11e8-9c93-deaee3d6c64b.png)

Via `php artisan tinker`:

```
=> [
     Nmap\Host {#828},
   ]
>>> print_r($hosts);
Array
(
    [0] => Nmap\Host Object
        (
            [addresses:Nmap\Host:private] => Array
                (
                    [95.143.172.170] => Nmap\Address Object
                        (
                            [address:Nmap\Address:private] => 95.143.172.170
                            [type:Nmap\Address:private] => ipv4
                            [vendor:Nmap\Address:private] =>
                        )

                )

            [state:Nmap\Host:private] => up
            [hostnames:Nmap\Host:private] => Array
                (
                    [0] => Nmap\Hostname Object
                        (
                            [name:Nmap\Hostname:private] => repat.de
                            [type:Nmap\Hostname:private] => user
                        )

                    [1] => Nmap\Hostname Object
                        (
                            [name:Nmap\Hostname:private] => aries.uberspace.de
                            [type:Nmap\Hostname:private] => PTR
                        )

                )

            [ports:Nmap\Host:private] => Array
                (
                    [0] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 22
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => ssh
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [1] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 80
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => http
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [2] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 110
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => pop3
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [3] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 113
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => ident
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [4] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 143
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => imap
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [5] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 443
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => https
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [6] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 587
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => submission
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [7] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 993
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => imaps
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [8] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 995
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => pop3s
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                )

        )

)
=> true
```

---

```php
$hosts = \Nmap\Nmap::create()->enableVerbose()->scan(['repat.de']);
```

In Laravel via the browser in a TestController:
```json
[
  {
    
  }
]
```

Via `php artisan tinker`:
```
Array
(
    [0] => Nmap\Host Object
        (
            [addresses:Nmap\Host:private] => Array
                (
                    [95.143.172.170] => Nmap\Address Object
                        (
                            [address:Nmap\Address:private] => 95.143.172.170
                            [type:Nmap\Address:private] => ipv4
                            [vendor:Nmap\Address:private] =>
                        )

                )

            [state:Nmap\Host:private] => up
            [hostnames:Nmap\Host:private] => Array
                (
                    [0] => Nmap\Hostname Object
                        (
                            [name:Nmap\Hostname:private] => repat.de
                            [type:Nmap\Hostname:private] => user
                        )

                    [1] => Nmap\Hostname Object
                        (
                            [name:Nmap\Hostname:private] => aries.uberspace.de
                            [type:Nmap\Hostname:private] => PTR
                        )

                )

            [ports:Nmap\Host:private] => Array
                (
                    [0] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 22
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => ssh
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [1] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 80
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => http
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [2] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 110
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => pop3
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [3] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 113
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => ident
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [4] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 143
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => imap
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [5] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 443
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => https
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [6] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 587
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => submission
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [7] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 993
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => imaps
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                    [8] => Nmap\Port Object
                        (
                            [number:Nmap\Port:private] => 995
                            [protocol:Nmap\Port:private] => tcp
                            [state:Nmap\Port:private] => open
                            [service:Nmap\Port:private] => Nmap\Service Object
                                (
                                    [name:Nmap\Service:private] => pop3s
                                    [product:Nmap\Service:private] =>
                                    [version:Nmap\Service:private] =>
                                )

                        )

                )

        )

)
=> true
```

**Comment by repat** [comment]
@m7mdcc please follow #19 . You're getting an empty array because nmap doesn't have sufficient rights. Please be careful with giving your webserver user root rights though

**Comment by m7mdcc** [comment]
Same version.


This code work for me


//Perform portscan
$nmap = new Nmap();
$nmap = Nmap::create()->scan([ $host['name'] ]);
$nmapScanHost = $nmap[0];
$nmapPorts = (Array)$nmapScanHost->getPorts();
$hostPorts = [];
foreach($nmapPorts as $port){
$hostPorts[] = ['number' => $port->getNumber(),'protocol' =>
$port->getProtocol(),'state' => $port->getState()];
}

**Comment by repat** [comment]
So the timeout thing seemed to have been the actual domain I was testing. I just installed `dev-master` and it works with Laravel 5.6.
